Extend swiss_public_transport integration

miaucl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
In the last few years, the swiss opendata public transport api has improved and a lot more parameters are available for requesting connections. This PR builds on the new version `0.4.0` of the `python_opendata_transport` [PyPI](https://pypi.org/project/python-opendata-transport/), which implements them. I have also contributed to that library, by implementing those parameters, with the goal to be able to integrate them into home assistant. 

Changes for the updated dependency: https://github.com/home-assistant-ecosystem/python-opendata-transport/blob/master/CHANGES.txt

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30213
- Related PR: https://github.com/home-assistant/core/pull/105648

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

joostlek reviewed (changes requested):
There may be no more updates to yaml based integrations if they connect to a device or service. They should be converted to Config flow first before receiving new features.

miaucl commented:
> There may be no more updates to yaml based integrations if they connect to a device or service. They should be converted to Config flow first before receiving new features.

Is that also the case for a public api? This integration allows to generate a variable list of sensors. I am new to this home assistant integration development and not sure how to approach that if I have to convert this integration in its entirety.

@fabaff Does it make sense to convert this integration to config flow?

miaucl committed:
extend with new parameters

add config flow

unit tests

yaml config import flow

change deprecation period and simply code

keep name for legacy yaml
- removing the name now would break current implementations
- it will be removed together with the deprectation of  yaml config flow

 improve error handling, simpler unique_id, cleanup

simplify issues for yaml import flow

improve typing and clean name handling

streamline unit tests
- happy path + errors
- mock opendata instead of aiohttp

parametrize unit tests

improve strings

add missing aborts

update coverage ignore

remove redundant test

Extend swiss_public_transport integration
- new version 0.4.0
- implement all parameters available by opendata api
- add offset parameter

add more parameters and test config flow

add unit tests for sensor

adapt silver medal constraints

- according to silver medal, one should log a warn when no connection
- according to silver medal, one should log a info when connection again
- use .availability property for connectivity information

joostlek commented:
I asked gjohansson to take a look at the config flow PR. After that, keep in mind that we still like small PRs (And you will as well as small PRs get reviewed quicker). I suggest bumping the library in a separate PR before adding new features. (And if you need to change stuff as part of that bump (for example you need to pass in an extra parameter), you can do that in that PR). If you want to discuss how to split up functionality, feel free to reach out on discord :)

miaucl commented:
> I asked gjohansson to take a look at the config flow PR. After that, keep in mind that we still like small PRs (And you will as well as small PRs get reviewed quicker). I suggest bumping the library in a separate PR before adding new features. (And if you need to change stuff as part of that bump (for example you need to pass in an extra parameter), you can do that in that PR). If you want to discuss how to split up functionality, feel free to reach out on discord :)

Thank you for the hint, I will go over this PR again anyway once the other one is merged. Will come back at you on discord if needed

miaucl commented:
Still waiting for review

miaucl commented:
keep it for now

miaucl commented:
keep it for now

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
